runtime/lua/vim/filetype.lua

```diff
@@ -2218,30 +2221,46 @@ local pattern = {
 -- luacheck: pop
 -- luacheck: pop
 
+--- Lookup table/cache for patterns
+--- @alias vim.filetype.pattern_cache { has_env: boolean, has_slash: boolean }
+--- @type table<string,vim.filetype.pattern_cache>
+local pattern_lookup = {}
+
+local function compare_by_priority(a, b)
+  return a[next(a)][2].priority > b[next(b)][2].priority
+end
+
 --- @param t vim.filetype.mapping
 --- @return vim.filetype.mapping[]
+--- @return vim.filetype.mapping[]
 local function sort_by_priority(t)
-  local sorted = {} --- @type vim.filetype.mapping[]
-  for k, v in pairs(t) do
-    local ft = type(v) == 'table' and v[1] or v
+  -- Separate patterns with non-negative and negative priority because they
+  -- will be processed separately
+  local pos = {} --- @type vim.filetype.mapping[]
+  local neg = {} --- @type vim.filetype.mapping[]
+  for pat, maptbl in pairs(t) do
+    local ft = type(maptbl) == 'table' and maptbl[1] or maptbl
     assert(
       type(ft) == 'string' or type(ft) == 'function',
       'Expected string or function for filetype'
     )
 
-    local opts = (type(v) == 'table' and type(v[2]) == 'table') and v[2] or {}
-    if not opts.priority then
-      opts.priority = 0
-    end
-    table.insert(sorted, { [k] = { ft, opts } })
+    -- Parse pattern for common data and cache it once
+    pattern_lookup[pat] = pattern_lookup[pat]
+      or { has_env = pat:find('%$%b{}') ~= nil, has_slash = pat:find('/') ~= nil }
```

Comment:
I'm guessing the point of this is to speed up `vim.filetype.add`?

Reply:
On second though, with the PR that adds the grouping, keeping a separate table is better as 99% of the time most of these patterns will never need to be processed.

There still might be merit to keeping this all in one table and lazy populating the fields.

Reply:
I see.

Is it too much to ask to leave it to some other time (some other contributor)? This whole refactor was to get the necessary pieces for #29660, not make it as perfect as it can be. And at least at the moment I'd prefer to do the reasonable minimum to get the performance gain of #29660 merged.

Reply:
I sniped a comment in before yours. We don't need to do this for now, and whatever we do, the parsing should be done lazily.